src/test/java/com/mongodb/hibernate/jdbc/MongoConnectionTests.java

```diff
@@ -16,56 +16,297 @@
+                verifyNoMoreInteractions(clientSession);
+            }
+
+            @Nested
+            class AutoCommitStateChangedTests {
+
+                @ParameterizedTest(name = "AutoCommit (true -> false): start transaction (successful: {0})")
+                @ValueSource(booleans = {true, false})
+                void testTringToStartTransactionWhenAutoCommitChangedToFalse(boolean successful) throws SQLException {
+                    // given
+                    assertTrue(mongoConnection.getAutoCommit());
+                    if (!successful) {
+                        doThrow(RuntimeException.class).when(clientSession).startTransaction();
+                    }
+
+                    // when && then
+                    if (successful) {
+                        mongoConnection.setAutoCommit(false);
+                    } else {
+                        assertThrows(SQLException.class, () -> mongoConnection.setAutoCommit(false));
+                    }
+                    verify(clientSession).startTransaction();
+                }
+
+                @ParameterizedTest(
+                        name = "AutoCommit changed (false -> true): commit existing transaction (successful: {0})")
+                @ValueSource(booleans = {true, false})
+                void testTryingToCommitExistingTransactionWhenAutoCommitChangedToTrue(boolean successful)
+                        throws SQLException {
+                    // given
+                    mongoConnection.setAutoCommit(false);
+                    verify(clientSession).startTransaction();
+
+                    assertFalse(mongoConnection.getAutoCommit());
+                    doReturn(true).when(clientSession).hasActiveTransaction();
+                    if (!successful) {
+                        doThrow(RuntimeException.class).when(clientSession).commitTransaction();
+                    }
+
+                    // when && then
+                    if (successful) {
+                        mongoConnection.setAutoCommit(true);
+                    } else {
+                        assertThrows(SQLException.class, () -> mongoConnection.setAutoCommit(true));
+                    }
+                    verify(clientSession).commitTransaction();
+                }
+
+                @Test
+                void testNoopWhenNoExistingTransactionAndAutoCommitChangedToTrue() throws SQLException {
+                    // given
+                    mongoConnection.setAutoCommit(false);
+                    verify(clientSession).startTransaction();
+                    mongoConnection.commit();
+                    verify(clientSession).commitTransaction();
+
+                    assertFalse(mongoConnection.getAutoCommit());
+
+                    // when
+                    mongoConnection.setAutoCommit(true);
+
+                    // then
+                    assertTrue(mongoConnection.getAutoCommit());
+                    verify(clientSession, times(2)).hasActiveTransaction();
```

Comment:
Given that we don't actually care about `hasActiveTransaction` being called twice, let's make the assertion less strict and, therefore, less prone to rising false alarms:

```suggestion
                    verify(clientSession, atLeast(1)).hasActiveTransaction();
```